### PR TITLE
Add Lampinen tests for differential_evolution

### PR DESCRIPTION
**Pull request by mdhaber**
_differentialevolution.py reference 5 (Lampinen) reports ten
benchmark problems. This commit adds the first as a test for
differential_evolution with constraints.

### CODE REVIEW

**Comment by andyfaff** [comment]
Just got back from holiday. Thanks for adding these. Do you have an idea for the timing of the tests? If they take more than a few seconds for each method, then some further work may need to be done. (e.g. reduce total number of iterations, func evals, etc).

The output from the paper may be different to that of scipy because:

1) the minimisation is stochastic
2) the total population sizes are different
3) the way that the population is initialised could be important
4) the `updating` kwd could play an important role here [best solution continually updated, or once per generation (traditional)].

File: scipy/optimize/tests/test__differential_evolution.py [hunk 1]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+
+        L = LinearConstraint(A[:5, :], -np.inf, b[:5])
+        L2 = LinearConstraint(A[5:6, :], -np.inf, b[5:6])
+        N = NonlinearConstraint(c1, -np.inf, b[6:8])
+        N2 = NonlinearConstraint(c2, -np.inf, b[8:9])
+        constraints = (L, N, L2, N2)
+        res = differential_evolution(f, bounds, constraints=constraints)
+
+        assert_allclose(res.x, x_opt, atol=1e-1)
+        assert_allclose(res.fun, f_opt, atol=0.5)
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: The `res.fun` (and res.x) tolerances for a lot of the tests are quite relaxed. What's the background for choosing the tolerances in these tests?

File: scipy/optimize/tests/test__differential_evolution.py [hunk 2]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+        L = LinearConstraint(A, -np.inf, b)
+
+        bounds = [(0, 1)]*9 + [(0, 100)]*3 + [(0, 1)]
+
+        # using parameters from [5]
+        res = differential_evolution(f, bounds, maxiter=4000, popsize=20,
+                                     mutation=0.9, recombination=0.9,
+                                     constraints=(L))
+
+        x_opt = (1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 1)
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: These `x_opt` are going to be right on the upper `bounds` limits. Did Lampinen do that (doesn't seem sensible to set bounds like that on purpose)?

File: scipy/optimize/tests/test__differential_evolution.py [hunk 3]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+
+        N = NonlinearConstraint(c1, 0, np.inf)
+        bounds = [(-10, 10)]*7
+        constraints = (N)
+        res = differential_evolution(f, bounds, maxiter=4000, popsize=20,
+                                     mutation=0.9, recombination=0.9,
+                                     constraints=constraints)
+
+        f_opt = 680.6300599487869
+        assert_allclose(res.fun, f_opt, atol=5)
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: f_opt tolerance is large. I would've hoped to get closer than this.

File: scipy/optimize/tests/test__differential_evolution.py [hunk 4]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+                    -3*(x[1]-2)**2 - 4*(x[2]-3)**2 - 2*x[3]**2 + 7*x[4] + 120,
+                    -x[1]**2 - 2*(x[2]-2)**2 + 2*x[1]*x[2] - 14*x[5] + 6*x[6],
+                    -5*x[1]**2 - 8*x[2] - (x[3]-6)**2 + 2*x[4] + 40,
+                    -0.5*(x[1]-8)**2 - 2*(x[2]-4)**2 - 3*x[5]**2 + x[6] + 30]
+
+        L = LinearConstraint(A, b, np.inf)
+        N = NonlinearConstraint(c1, 0, np.inf)
+        bounds = [(-10, 10)]*10
+        constraints = (L, N)
+        res = differential_evolution(f, bounds, maxiter=5000, popsize=35, mutation=0.9, recombination=0.9, constraints=constraints)
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: maxiter is quite large (for all examples). Normally with an unconstrained minimisation 100-200 is ok. I wonder if applying constraints mean that more iterations are needed.

File: scipy/optimize/tests/test__differential_evolution.py [hunk 5]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+        A = A[1:, 1:]
+
+        b = np.array([10, 0, 0, 10, 0, 0, 10, 0, 0])
+
+        L = LinearConstraint(A, -np.inf, b)
+
+        bounds = [(0, 1)]*9 + [(0, 100)]*3 + [(0, 1)]
+
+        # using parameters from [5]
+        res = differential_evolution(f, bounds, maxiter=4000, popsize=20,
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: If following Lampinen exactly is important then `updating='deferred'` probably needs to be used. Otherwise I'd stick to the default `differential_evolution` keyword arguments.

File: scipy/optimize/tests/test__differential_evolution.py [hunk 6]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+                                     constraints=(L))
+
+        x_opt = (1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 1)
+        f_opt = -15
+
+        assert_allclose(res.x, x_opt, atol=1e-1)
+        assert_allclose(res.fun, f_opt, atol=0.5)
+        assert_(np.all(A@res.x <= b))
+        assert_(np.all(res.x >= np.array(bounds)[:, 0]))
+        assert_(np.all(res.x <= np.array(bounds)[:, 1]))
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: Where appropriated can we check that `res.maxcv` is zero for all the tests (to show that the constraint mechanism works)?

File: scipy/optimize/tests/test__differential_evolution.py [hunk 7]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+                    ]
+
+        N = NonlinearConstraint(c1, [0, 90, 20], [92, 110, 25])
+
+        bounds = [(78, 102), (33, 45)]+[(27, 45)]*3
+        constraints = (N)
+        res = differential_evolution(f, bounds, maxiter=2000, popsize=15,
+                                     mutation=0.9, recombination=0.9,
+                                     constraints=constraints)
+        x_opt = (78.0, 33.0, 29.995, 45.0, 36.776)
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: first two `x_opt`s are right on the `bounds` here. Is that correct?

File: scipy/optimize/tests/test__differential_evolution.py [hunk 8]
```diff
@@ -666,3 +667,256 @@ def constr_f(x):
...
+        b = np.array([10, 0, 0, 10, 0, 0, 10, 0, 0])
+
+        L = LinearConstraint(A, -np.inf, b)
+
+        bounds = [(0, 1)]*9 + [(0, 100)]*3 + [(0, 1)]
+
+        # using parameters from [5]
+        res = differential_evolution(f, bounds, maxiter=4000, popsize=20,
+                                     mutation=0.9, recombination=0.9,
+                                     constraints=(L))
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: Is it worth setting a seed for all the tests? For a stochastic minimiser it can be the case that a certain proportion fail if the test is hard. This would lead to intermittent test suite fails.

File: scipy/optimize/tests/test__differential_evolution.py [hunk 9]
```diff
@@ -920,3 +920,30 @@ def c1(x):
...
+
+        def c1(x):
+            x = np.hstack(([0], x))  # 1-indexed to match reference
+            return [x[2] - x[1]**2]
+
+        N = NonlinearConstraint(c1, [-.001], [0.001])
+
+        bounds = [(-1, 1)]*2
+        constraints = (N)
+        res = differential_evolution(f, bounds, maxiter=1000, popsize=30,
```
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py]: As an example for the different population sizes employed - this example will use a total population size of 60 (`len(x) * popsize`). In comparison Lampinen uses a population size of 30 in the paper. All other things being equal this would mean it would take twice as long for each iteration.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 2 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: It's likely the other way around. The bounds were set, and the optimal solution under the constraints has many bounds active.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 6 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: You mean to show that whatever is calculating `maxcv` works? Clearly some sort of 'constraint mechanism' is working, as all these manual constraint checks are passing.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 1 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: I don't know what sort of accuracy you expect, so I just set them wide enough so that they would pass (based on a few runs).

File: scipy/optimize/tests/test__differential_evolution.py (hunk 4 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: I set these based on Table II.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 7 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: It is very common for constraints to be active at an optimum. It (typically) indicates that the objective function would be even better but the bound got in the way.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 9 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: Ok. That is easily fixed. I've never used `differential_evolution` before this, and I didn't study the difference in terminology between the paper and the options.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 3 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: You might check out the PR or copy-paste the tests so you can investigate. That's what these tests are for, after all: to reveal any unexpected behavior.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 8 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: Yes.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 5 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: I don't think it's important. I only attempted to follow Lampinen's parameters because I wasn't observing the same accuracy with default parameters. It didn't get much better after setting the parameters, so I think it's fine to use defaults.

**Comment by mdhaber** [comment]
I just wanted to help out because it was unfortunate that I couldn't approve the PR for 1.3.  Since it's unlikely to get a qualified, deep technical review of the code, extensive testing on real benchmark problems like this is an important substitute.

Based on the results, I would approve the PR (once we add the fastest of these problems to the test suite and the rest as benchmarks). `differential_evolution` seems to reliably find the vicinity of the global optimum on all problems tested - close enough to use as a guess for a local minimizer, which typically gets stuck in other local minima otherwise. That's really valuable, and personally, that's all I expected. (I've only ever used local minimizers before, though, so maybe my expectations for global optimizers are too low?)

However, as you note, the speed and accuracy may not be as hoped, so you're welcome to run these tests on locally and investigate the results. I'd be very curious whether you can replicate the results of Lampinen with the appropriate choice of options.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 5 above)
Comment (andyfaff) [reply scipy/optimize/tests/test__differential_evolution.py]: defaults would go faster.

**Comment by mdhaber** [comment]
I've read the paper more carefully. I'll adjust the tests as you suggested. I'll also try to review the code. You're right that it should be getting closer to the optimal objective value. Even in the worst of the 1000 runs, Lampinen reports an objective within .001 of the minimum for 8 of the 10 problems.

**Comment by mdhaber** [comment]
I've gotten Dual Annealing failures in Trivis-CI a few times now. Do you know why that happens?

**Comment by andyfaff** [comment]
I'm not sure about that. It seems to be fine on the macOS entry, but fails on Linux. The fail is just outside the tolerance.
In any case no changes have been made to the dual annealing code, nor SLSQP. `dual_annealing` is also a stochastic minimiser, so my educated guess that it's something to do with its random number usage. That particular test doesn't set a seed, so relies on the immediate state of the numpy RNG - the state is apparently such that the example fails. Thus, the fix is either:

1. setting a seed for the test which will result in success, or
2. relax the test tolerance slightly.

My vote is for option 1, but that's a separate PR. For stochastic minimisers it's always a good idea to set the seed to ensure reproducibility in testing.

**Comment by andyfaff** [comment]
I've submitted a PR to set a seed. Let's see if that fixes things.

File: scipy/optimize/tests/test__differential_evolution.py [hunk 10]
```diff
@@ -703,6 +703,7 @@ def f(x):
         x_opt = (1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 1)
         f_opt = -15
 
+        assert_allclose(f(x_opt), f_opt)
```
Comment (mdhaber) [inline scipy/optimize/tests/test__differential_evolution.py]: Did you mean to do this? f_opt, x_opt, and f are all hardcoded. This checks that test itself was coded consistently but doesn't test `differential_evolution`, right?

File: scipy/optimize/tests/test__differential_evolution.py [hunk 11]
```diff
@@ -756,6 +757,9 @@ def c1(x):
                                      constraints=constraints)
 
         f_opt = 680.6300599487869
+        x_opt = (2.330499, 1.951372, -0.4775414, 4.365726, -0.6244870, 1.038131, 1.594227)
```
Comment (mdhaber) [inline scipy/optimize/tests/test__differential_evolution.py]: I think the reason I didn't include this `x_opt` before was that there was too much variability in the solution to make an `assert_allclose(res.x, x_opt)` meaningful (the tolerances would have to be too big). But hopefully we can improve the accuracy, so it's good to add this.

**Comment by andyfaff** [comment]
I have a pretty good idea for why the examples aren't getting closer to the known optimum. The solver stops when the convergence criterion is satisfied:

    def converged(self):
        """
        Return True if the solver has converged.
        """
        return (np.std(self.population_energies) <=
                self.atol +
                self.tol * np.abs(np.mean(self.population_energies)))

in plain English, when the standard deviation of the population energies divided by its mean is less than some tolerance (currently 0.01, 1% variability in population energies) then the solving process stops. Thus, for example L1 even if you ask for `maxiter=4000` it may stop after much less (e.g. couple of hundred). With an unconstrained  problem the idea is to get the 'polishing' process to take you the rest of the way because it's more efficient in terms of function evaluations, and it's reasonably probable that you're in the global minimum. For constrained problems the smaller number of iterations isn't sufficient to get really close to the known solution.

With the publicly exposed `differential_evolution` function there is no clear way (yet) of asking for a minimum/certain number of iterations to take place. One can decrease the tolerance which does require more iterations to take place, and it does get you a lot closer to the known solution, but that's not quite the same thing.

With the private `DifferentialEvolutionSolver` class one can use it as an iterator:

    solver = DifferentialEvolutionSolver(f, bounds)
    for i in range(1500):
        _ = next(solver)
    
    res = solver.solve()

When I used this approach for example L2 I can get within 2e-4 absolute for `f_opt` and within 1.1e-3 absolute for `x_opt` after 1500 iterations. If one leaves it up to the normal criterion for stopping then it only runs 112 iterations, getting within 5.74 absolute for `f_opt`.

In my experience for unconstrained `differential_evolution` the main solve gets you pretty much all the way to the solution and the polisher does very little. For a constrained minimisation problem the default convergence criterion activates whilst the solver is still able to do more. I'm guessing this is because as  feasible solutions start to appear in the population they rapidly propagate which may lead to an overly premature convergence.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 10 above)
Comment (andyfaff) [reply scipy/optimize/tests/test__differential_evolution.py]: One of the costly lessons I've experienced in life is that it's always good to do this kind of check to ensure that the cost function is being calculated correctly.

**Comment by andyfaff** [comment]
In addition to my last comment; I tried doing a `trust-constr` polish on example L2 after the normal criterion for stopping (112 iterations), it gets to the best solution pretty quickly.

So perhaps it's best to use the `trust-constr` polish option for the solver, which is what you were suggesting anyway.

A.

**Comment by andyfaff** [comment]
Matt, thanks for all your work on this. I'm going to merge this into my branch now, as I'd like to do further work on this.

File: scipy/optimize/tests/test__differential_evolution.py (hunk 10 above)
Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: Ok. I checked this for all the problems, of course, but if you're concerned that the test might inadvertently be modified I understand.

Comment (andyfaff) [reply scipy/optimize/tests/test__differential_evolution.py]: When I wrote the >120 benchmark functions for examining global optimisers I used the x_opt and f_opt from original source papers. More frequently than not `f(x_opt) != f_opt`. Indeed, a lot of the time f_opt wasn't even the most optimal.

Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py]: Huh! Well, all of these ones seemed to match, and DE seemed to approach them. I manually polished most of them with trust_constr and I think it got slightly better numbers on one, but the improvement was small compared to the difference between the variation in DE's solution, so I didn't really take note.

**Comment by mdhaber** [comment]
Good to know that more iterations does the trick. 

I'll try this at some point to see if DE gives comparable results to the paper for most of the problems when forced to run for the same number of iterations.
